### PR TITLE
feat: Fee estimate RPC

### PR DESCRIPTION
**Pull request by jjyr**
This PR adds a new RPC [estimate_fee_rate](https://github.com/nervosnetwork/ckb/pull/1659/files#diff-622e6d119ac5d43f7eb41cb596159f9fR907). It takes the basic idea from bitcoin's [estimatesmartfee](https://bitcoincore.org/en/doc/0.16.0/rpc/util/estimatesmartfee/), however, we ignore the magic numbers and tricks from the original code.

We estimate the tx fee rate by track txs that entered tx pool. See details https://github.com/nervosnetwork/ckb/pull/1659/files#diff-ff03764a87b23e747dadc645fcf8df8bR21

### CODE REVIEW

**Comment by jjyr** [comment]
> Need some integration tests for `estimate_fee_rate` rpc

There is a basic test case in the rpc_test, what the other cases do you want?

File: util/fee-estimator/src/tx_confirm_stat.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+/// So we get a group of samples which includes txs count, average fee rate and confirmed blocks, etc.
+/// For estimate, we loop through each bucket, calculate the confirmed txs rate, until meet the required_confirm_rate.
+#[derive(Clone)]
+pub struct TxConfirmStat {
+    /// per bucket stat
+    bucket_stats: Vec<BucketStat>,
+    /// bucket upper bound fee_rate => bucket index
+    fee_rate_to_bucket: BTreeMap<FeeRate, usize>,
+    /// confirm_blocks => bucket index => confirmed txs count
+    confirm_blocks_to_confirmed_txs: Vec<Vec<f64>>,
```
Comment (quake) [inline util/fee-estimator/src/tx_confirm_stat.rs]: Should it be `Vec<Vec<u64>>` since it's a counter?

**Comment by quake** [comment]
One more question:  the `Estimator#track_tx` method should be invoked when a tx entered pool, but I can't find related code in the `tx-pool` module, could you double check it?

File: util/fee-estimator/src/tx_confirm_stat.rs (hunk 1 above)
Comment (jjyr) [reply util/fee-estimator/src/tx_confirm_stat.rs]: See decay

File: tx-pool/src/service.rs [hunk 2]
```diff
@@ -227,6 +232,39 @@ impl TxPoolController {
...
+        let (responder, response) = crossbeam_channel::bounded(1);
+        let request = Request::call((tx_hash, fee_rate, height), responder);
+        sender.try_send(Message::EstimatorTrackTx(request))?;
+        response.recv().map_err(Into::into)
+    }
+
+    pub fn estimator_process_block(
+        &self,
+        height: u64,
+        txs: impl Iterator<Item = Byte32>,
```
[the remark this answers is not in the document]
Comment (jjyr) [reply tx-pool/src/service.rs]: This is a local variable, apparently everyone knows it's tx hash, there's no chance to confuse this.

File: util/fee-estimator/src/estimator.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        // For simpfy, we assume chain reorg will not effect tx fee.
+        if height <= self.best_height {
+            return;
+        }
+        self.best_height = height;
+        // update tx confirm stat
+        self.tx_confirm_stat.move_track_window(height);
+        self.tx_confirm_stat.decay();
+        let processed_txs = txs.filter(|tx| self.process_block_tx(height, tx)).count();
+        if self.start_height == 0 && processed_txs > 0 {
```
[the remark this answers is not in the document]
Comment (jjyr) [reply util/fee-estimator/src/estimator.rs]: Mainly used for dialog, I'll add a debug log for this.

File: util/fee-estimator/src/tx_confirm_stat.rs (hunk 1 above)
Comment (jjyr) [reply util/fee-estimator/src/tx_confirm_stat.rs]: We use decay to reduce the effect from old records, so it be an f64 can makes things easier.

**Comment by jjyr** [comment]
> One more question: the `Estimator#track_tx` method should be invoked when a tx entered pool, but I can't find related code in the `tx-pool` module, could you double check it?

Fixed, It's an error during rebase.

**Comment by jjyr** [comment]
Rebase during `Cargo.lock` conflict, please re-approve @quake @zhangsoledad

**Comment by jjyr** [comment]
bors r=quake, zhangsoledad

**Comment by jjyr** [comment]
bors r=quake, zhangsoledad

**Comment by jjyr** [comment]
bors r=quake, zhangsoledad

**Pull request merged**